externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/operation/ExecutePython.scala

```diff
@@ -0,0 +1,238 @@
+import org.apache.spark.api.python.KyuubiPythonGatewayServer
+import org.apache.spark.sql.Row
+import org.apache.spark.sql.types.StructType
+
+import org.apache.kyuubi.Logging
+import org.apache.kyuubi.operation.ArrayFetchIterator
+import org.apache.kyuubi.session.Session
+
+class ExecutePython(
+    session: Session,
+    override val statement: String,
+    worker: SessionPythonWorker) extends SparkOperation(session) {
+
+  override protected def resultSchema: StructType = {
+    if (result == null || result.schema.isEmpty) {
+      new StructType().add("output", "string")
+        .add("status", "string")
+        .add("ename", "string")
+        .add("evalue", "string")
+        .add("traceback", "array<string>")
+    } else {
+      result.schema
+    }
+  }
+
+  override protected def runInternal(): Unit = {
+    val response = worker.runCode(statement)
+    val output = response.map(_.content.getOutput()).getOrElse("")
+    val status = response.map(_.content.status).getOrElse("UNKNOWN_STATUS")
+    val ename = response.map(_.content.getEname()).getOrElse("")
+    val evalue = response.map(_.content.getEvalue()).getOrElse("")
+    val traceback = response.map(_.content.getTraceback()).getOrElse(Array.empty)
+    iter =
+      new ArrayFetchIterator[Row](Array(Row(output, status, ename, evalue, Row(traceback: _*))))
+  }
+
+}
+
+case class SessionPythonWorker(
+    errorReader: Thread,
+    pythonWorkerMonitor: Thread,
+    workerProcess: Process) {
+  private val stdin: PrintWriter = new PrintWriter(workerProcess.getOutputStream)
+  private val stdout: BufferedReader =
+    new BufferedReader(new InputStreamReader(workerProcess.getInputStream), 1)
+
+  def runCode(code: String): Option[PythonReponse] = {
+    val input = ExecutePython.toJson(Map("code" -> code, "cmd" -> "run_code"))
+    // scalastyle:off println
+    stdin.println(input)
+    // scalastyle:on
+    stdin.flush()
+    Option(stdout.readLine())
+      .map(ExecutePython.fromJson[PythonReponse](_))
+  }
+
+  def close(): Unit = {
+    val exitCmd = ExecutePython.toJson(Map("cmd" -> "exit_worker"))
+    // scalastyle:off println
+    stdin.println(exitCmd)
+    // scalastyle:on
+    stdin.flush()
+    stdin.close()
+    stdout.close()
+    errorReader.interrupt()
+    pythonWorkerMonitor.interrupt()
+    workerProcess.destroy()
+  }
+}
+
+object ExecutePython extends Logging {
+
+  // TODO:(fchen) get from conf
+  val pythonExec =
+    sys.env.getOrElse("PYSPARK_PYTHON", sys.env.getOrElse("PYSPARK_DRIVER_PYTHON", "python3"))
+  private val isPythonGatewayStart = new AtomicBoolean(false)
+  val kyuubiPythonPath = Files.createTempDirectory("")
+  def init(): Unit = {
+    if (!isPythonGatewayStart.get()) {
+      synchronized {
+        KyuubiPythonGatewayServer.start()
+        writeTempPyFile(kyuubiPythonPath, "execute_python.py")
+        writeTempPyFile(kyuubiPythonPath, "kyuubi_util.py")
+        isPythonGatewayStart.set(true)
+      }
+    }
+  }
```

Comment:
```suggestion
private val kyuubiPythonPath = Files.createTempDirectory("")
  lazy val init: Unit = {
    KyuubiPythonGatewayServer.start()
    writeTempPyFile(kyuubiPythonPath, "execute_python.py")
    writeTempPyFile(kyuubiPythonPath, "kyuubi_util.py")
  }
```

simply to use `lazy val` to lock the object instance instead.

Reply:
How does the `lazy val` expression lock the object instance?

Reply:
`lazy val` in Scala basically equivalent to auto added method in compile time as `synchronized (this) +  double check empty` in Java

Reply:
> How does the `lazy val` expression lock the object instance?

FYI: https://www.baeldung.com/scala/lazy-val

Reply:
> How about simply use lazy val to lock the object instance instead.

I think `double-check` style is more readable than `lazy val`, WDYT? cc @pan3793

Reply:
But more readness burden for one-time usage, and introduced more memory footprint for this single check.

Reply:
either is fine to me.